### PR TITLE
[tests-only] Feature to skip and debug test suites in CI

### PR DESCRIPTION
**Pull request by saw-jan**
## Description
Adds the ability to skip pipelines and debug specific test suites. That will make it easier when debugging tests in drone CI.

## Related Issue
- part of https://github.com/owncloud/QA/issues/670

## How Has This Been Tested?
- test environment:

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: .drone.star [hunk 1]
```diff
@@ -527,8 +547,35 @@ def coreApiTests(ctx, part_number = 1, number_of_parts = 1, storage = "owncloud"
...
+def apiTests(ctx, pipelines):
+    partsToSkip = config["apiTests"]["skipExceptParts"]
+    for runPart in range(1, config["apiTests"]["numberOfParts"] + 1):
+        if len(partsToSkip) != 0:
+            if runPart in partsToSkip:
+                pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "owncloud"))
+                pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "ocis"))
+        else:
+            pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "owncloud"))
+            pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "ocis"))
```
Comment (jasson99) [inline .drone.star:559]: Aren't these exactly the same?

Comment (saw-jan) [reply .drone.star:559]: Nope,
it conditionally appends pipelines.
If I want to run only part 5 of 10 (numberOfParts) with `skipExceptParts: [5]` then `apiTests()` will append only that pipeline.
Else the whole pipelines (1-10 parts) are appended.

File: .drone.star [hunk 2]
```diff
@@ -527,8 +547,35 @@ def coreApiTests(ctx, part_number = 1, number_of_parts = 1, storage = "owncloud"
         "volumes": [pipelineVolumeOC10Tests],
     }
 
+def apiTests(ctx, pipelines):
+    partsToSkip = config["apiTests"]["skipExceptParts"]
```
Comment (saw-jan) [inline .drone.star:551]: That one should be
```suggestion
    partsNotToSkip = config["apiTests"]["skipExceptParts"]
```

Comment (jasson99) [reply .drone.star]: more clear now :+1:

File: .drone.star [hunk 3]
```diff
@@ -247,20 +260,27 @@ def testOcisModules(ctx):
...
+            localApiTests(ctx, "ocis", "apiBugDemonstration"),
+            localApiTests(ctx, "owncloud", "apiAccountsHashDifficulty", "default"),
+            localApiTests(ctx, "ocis", "apiAccountsHashDifficulty", "default"),
+        ]
 
-    for runPart in range(1, config["apiTests"]["numberOfParts"] + 1):
-        pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "owncloud"))
-        pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "ocis"))
+    if "skip" not in config["apiTests"] or not config["apiTests"]["skip"]:
+        pipelines = apiTests(ctx, pipelines)
```
Comment (phil-davis) [inline .drone.star:273]:
```suggestion
        pipelines += apiTests(ctx, pipelines)
```

File: .drone.star [hunk 4]
```diff
@@ -527,8 +547,33 @@ def coreApiTests(ctx, part_number = 1, number_of_parts = 1, storage = "owncloud"
...
+def apiTests(ctx, pipelines):
+    debugParts = config["apiTests"]["skipExceptParts"]
+    for runPart in range(1, config["apiTests"]["numberOfParts"] + 1):
+        if len(debugParts) != 0:
+            if runPart in debugParts:
+                pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "owncloud"))
+                pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "ocis"))
+        else:
+            pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "owncloud"))
+            pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "ocis"))
```
Comment (phil-davis) [inline .drone.star:559]:
```suggestion
    debugPartsEnabled = (len(debugParts) != 0)
    for runPart in range(1, config["apiTests"]["numberOfParts"] + 1):
        if (!debugPartsEnabled or (debugPartsEnabled and runPart in debugParts):
                pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "owncloud"))
                pipelines.append(coreApiTests(ctx, runPart, config["apiTests"]["numberOfParts"], "ocis"))
```
We could avoid repeating the `pipelines.append` lines by condensing the logic - something like this.

**Review by Talank — COMMENTED** [review]
specify this feature in documentation as well.

**Comment by phil-davis** [comment]
> specify this feature in documentation as well.

I raised docs issue https://github.com/owncloud/docs/issues/3549 - the documentation will be common for all the different repos that we are doing this in.

**Review by phil-davis — APPROVED** [review]

**Pull request merged**